src/string_radix_sort.jl

```diff
@@ -0,0 +1,233 @@
+        end
+        return sortperm_radixsort(o.data, order = o.order)
+    else
+        sort!(svec, lo, hi, StringRadixSort, o)
+    end
+end
+
+function sort!(svec::AbstractVector{String}, lo::Int, hi::Int, ::StringRadixSortAlg, o::O) where O <: Union{ForwardOrdering, ReverseOrdering, LexicographicOrdering, Perm}
+    #  Input checking
+    if isa(o, Perm)
+        if eltype(o.data) != String
+            throw(ArgumentError("Cannot use StringRadixSort on type $(eltype(o.data))"))
+        end
+        o = o.order
+        return
+    end
+    
+    if lo >= hi;  return svec;  end
+
+    # find the maximum string length
+    lens = reduce((x,y) -> max(x,sizeof(y)),0, svec)
+    skipbytes = lens
+    while lens > 0
+       if lens > 8
+            skipbytes = max(0, skipbytes - 16)
+            if o == Reverse
+                sorttwo!(.~load_bits.(UInt128, svec, skipbytes), svec)
+            else
+                sorttwo!(load_bits.(UInt128, svec, skipbytes), svec)
+            end
+            lens -= 16
+        elseif lens > 4
+            skipbytes = max(0, skipbytes - 8)
+            if o == Reverse
+                sorttwo!(.~load_bits.(UInt64, svec, skipbytes), svec)
+            else
+                sorttwo!(load_bits.(UInt64, svec, skipbytes), svec)
+            end
+            lens -= 8
+        else
+            skipbytes = max(0, skipbytes - 4)
+            if o == Reverse
+                sorttwo!(.~load_bits.(UInt32, svec, skipbytes), svec)
+            else
+                sorttwo!(load_bits.(UInt32, svec, skipbytes), svec)
+            end
+            lens -= 4
+        end
+    end
+    svec
+end
+
+"""
+    sorttwo!(vs, index)
+
+Sort both the `vs` and reorder `index` at the same. This allows for faster sortperm
+for radix sort.
+"""
+function sorttwo!(vs::AbstractVector{T}, index, lo::Int = 1, hi::Int=length(vs)) where T
+    # Input checking
+    if lo >= hi;  return (vs, index);  end
+
+    # Make sure we're sorting a bits type
+    # T = Base.Order.ordtype(o, vs)
+    if !isbits(T)
+        error("Radix sort only sorts bits types (got $T)")
+    end
+    o = Forward
+
+    # Init
+    iters = ceil(Integer, sizeof(T)*8/RADIX_SIZE)
+    bin = zeros(UInt32, 2^RADIX_SIZE, iters)
+    if lo > 1;  bin[1,:] = lo-1;  end
+
+    # Histogram for each element, radix
+    for i = lo:hi
+        v = uint_mapping(o, vs[i])
+        for j = 1:iters
+            idx = Int((v >> (j-1)*RADIX_SIZE) & RADIX_MASK) + 1
+            @inbounds bin[idx,j] += 1
+        end
+    end
+
+    # Sort!
+    swaps = 0
+    len = hi-lo+1
+
+    index1 = similar(index)
+    ts=similar(vs)
+    for j = 1:iters
+        # Unroll first data iteration, check for degenerate case
+        v = uint_mapping(o, vs[hi])
+        idx = Int((v >> (j-1)*RADIX_SIZE) & RADIX_MASK) + 1
+
+        # are all values the same at this radix?
+        if bin[idx,j] == len;  continue;  end
+
+        cbin = cumsum(bin[:,j])
```

Comment:
Maybe use `cumsum!` with a preallocated array.  Note that `bin[:,j]` also allocates a copy.  You might want to just write a loop to do the sum with no copies.

Reply:
done